sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -170,21 +170,35 @@ class Analyzer(catalog: Catalog,
    * Replaces [[UnresolvedRelation]]s with concrete relations from the catalog.
    */
   object ResolveRelations extends Rule[LogicalPlan] {
-    def getTable(u: UnresolvedRelation) = {
+    def getTable(u: UnresolvedRelation, cteRelations: Map[String, LogicalPlan]) = {
       try {
-        catalog.lookupRelation(u.tableIdentifier, u.alias)
+          // In hive, if there is same table name in database and CTE definition,
+          // hive will use the table in database, not the CTE one.
+          // Taking into account the reasonableness and the implementation complexity,
+          // here use the CTE definition first, check table name only and ignore database name
+          cteRelations.get(u.tableIdentifier.last)
```

Comment:
I don't think we should just drop a database name if one is specified.  Instead can we check to make sure `u.tableIdentifier.size == 1`, and only then look for CTEs?  That way you can use a fully qualified name to override a CTE.

Then I agree that it seems better to check CTEs first and only if one does not exist check the catalog.